spec/machine/AARCH64/MachineOps.thy

```diff
@@ -178,31 +179,28 @@ definition getFPUState :: "fpu_state user_monad" where
 definition setFPUState :: "fpu_state \<Rightarrow> unit user_monad" where
   "setFPUState fc \<equiv> modify (\<lambda>s. UserContext fc (user_regs s))"
 
-consts' nativeThreadUsingFPU_impl :: "machine_word \<Rightarrow> unit machine_rest_monad"
-consts' nativeThreadUsingFPU_val :: "machine_state \<Rightarrow> bool"
-definition nativeThreadUsingFPU :: "machine_word \<Rightarrow> bool machine_monad" where
-  "nativeThreadUsingFPU thread_ptr \<equiv> do
-       machine_op_lift (nativeThreadUsingFPU_impl thread_ptr);
-       gets nativeThreadUsingFPU_val
-  od"
+consts' readFpuState_val :: "machine_state_rest \<Rightarrow> fpu_state"
+definition readFpuState :: "fpu_state machine_monad" where
+  "readFpuState \<equiv>  do
+     state_assert fpu_enabled;
+     machine_rest_lift $ gets readFpuState_val
+   od"
+
+consts' writeFpuState_impl :: "fpu_state \<Rightarrow> unit machine_rest_monad"
+definition writeFpuState :: "fpu_state \<Rightarrow> unit machine_monad" where
+  "writeFpuState val \<equiv>  do
+     state_assert fpu_enabled;
+     machine_op_lift $ writeFpuState_impl val
+   od"
+
+definition enableFpu :: "unit machine_monad" where
+  "enableFpu \<equiv> modify (\<lambda>s. s\<lparr>fpu_enabled := True \<rparr>)"
```

Comment:
Left over from internal review:

@Xaphiosis 

> the (lack of) link between `enableFpu` and `enableFpuEL01` confuses me... there's no machine op for disabling the FPU? Are `enableFpu` and `enableFpuEL01` called separately by the C, or is this only the abstract spec model?

@corlewis 

> I was confused also. In the C, `enableFpu` calls either `disableTrapFpu` or `enableFpuEL01`, depending on whether it is a configuration with hypervisor support (see https://github.com/seL4/seL4/blob/85aa104eb4d527bc9f0aac28f3e9c3a84eed1d7a/include/arch/arm/arch/64/mode/machine/fpu.h#L127). I was originally going to just have `enableFpu` in the spec before realising that `enableFpuEL01` is called directly from `vcpu_disable` (https://github.com/seL4/seL4/blob/85aa104eb4d527bc9f0aac28f3e9c3a84eed1d7a/include/arch/arm/armv/armv8-a/64/armv/vcpu.h#L702)

@Xaphiosis 

> That seems odd, because it is unconditionally enabling FPU for native threads with no VCPU on hyp configurations. Doesn't that effectively ignore the native thread's configured flags?

Reply:
To answer this earlier question, our confusion was due to the fact that on AArch64 with hypervisor support there are two registers that control whether the FPU is enabled. The first, `CPACR_EL1`, is generally controlled by the VM and is used to enable or disable the FPU for virtualised threads. This register is the one being enabled by `enableFpuEL01`, which seL4 only calls when disabling the VCPU. For native threads seL4 instead uses `CPTR_EL2`, which it modifies with `enableFpu` and `disableFpu`.